test/test_solve_powerflow.jl

```diff
@@ -907,3 +618,145 @@ end
+
+    pf = ACPowerFlow(;
+        generator_slack_participation_factors = generator_slack_participation_factors,
+    )
+
+    data = PowerFlowData(pf, sys; check_connectivity = false)
+    data_original_bus_power = copy(data.bus_activepower_injection[:, 1])
+    bus_numbers = get_bus_numbers(sys)
+    original_bus_power, original_gen_power = _system_generation_power(sys, bus_numbers)
+
+    solve_powerflow!(pf, sys; check_connectivity = false)
+
+    if mode == :same
+        @test isapprox(get_active_power(s1), get_active_power(g1), rtol = 0, atol = 1e-6)
+        @test isapprox(get_active_power(s2), get_active_power(g2), rtol = 0, atol = 1e-6)
+
+        @test isapprox(get_active_power(s1), get_active_power(s2), rtol = 0, atol = 1e-6)
+        @test isapprox(get_active_power(g1), get_active_power(g2), rtol = 0, atol = 1e-6)
+    end
+
+    # tol of 1e-3 due to losses
+    @test isapprox(
+        get_active_power(ld1),
+        get_active_power(s1) + get_active_power(g1),
+        rtol = 0,
+        atol = 1e-3,
+    )
+    @test isapprox(
+        get_active_power(ld2),
+        get_active_power(s2) + get_active_power(g2),
+        rtol = 0,
+        atol = 1e-3,
+    )
+
+    solve_powerflow!(data; pf = pf)
+
+    @test isapprox(
+        data.bus_activepower_injection[data.bus_lookup[get_number(b1)], 1],
+        get_active_power(s1),
+        rtol = 0,
+        atol = 1e-6,
+    )
+    @test isapprox(
+        data.bus_activepower_injection[data.bus_lookup[get_number(b2)], 1],
+        get_active_power(g1),
+        rtol = 0,
+        atol = 1e-6,
+    )
+    @test isapprox(
+        data.bus_activepower_injection[data.bus_lookup[get_number(b4)], 1],
+        get_active_power(s2),
+        rtol = 0,
+        atol = 1e-6,
+    )
+    @test isapprox(
+        data.bus_activepower_injection[data.bus_lookup[get_number(b5)], 1],
+        get_active_power(g2),
+        rtol = 0,
+        atol = 1e-6,
+    )
+
+    bus_slack_participation_factors = zeros(Float64, length(bus_numbers))
+    for bn in bus_numbers
+        bus = get_bus(sys, bn)
+        idx = data.bus_lookup[bn]
+        data.bus_type[idx, 1] == ACBusTypes.PQ && continue
+        bus_slack_participation_factors[idx] = data.bus_slack_participation_factors[idx, 1]
+    end
+
+    # needed for the test implementation when data.bus_slack_participation_factors is compared to bus_slack_participation_factors
+    generator_slack_participation_factors2 = Dict(
+        (typeof(x), get_name(x)) => f for
+        (x, f) in zip(devices, factors) if get_bustype(get_bus(x)) != ACBusTypes.PQ
+    )
+    pf2 = ACPowerFlow(;
+        generator_slack_participation_factors = generator_slack_participation_factors2,
+    )
+
+    _check_ds_pf(
+        pf2,
+        sys,
+        bus_slack_participation_factors,
+        bus_numbers,
+        original_bus_power,
+        original_gen_power,
+        data_original_bus_power;
+        check_connectivity = false,
+    )
+end
```

Comment:
Test the case of multiple `ACBusTypes.REF` buses in a given subnetwork, also test that it errors properly when there are zero `ACBusTypes.REF` buses in a given subnetwork